src/app.js

```diff
@@ -215,48 +217,134 @@ export default function (app) {
       var realValue = element[name]
 
       if (name === "onUpdate") {
-        defer(value, element)
+          defer(value, element)
 
       } else if (value !== oldValue || realValue !== value) {
         setElementData(element, name, value, oldValue)
       }
     }
   }
 
-  function patch(parent, element, oldNode, node) {
-    if (oldNode === undefined) {
-      element = parent.appendChild(createElementFrom(node))
-
-    } else if (node === undefined) {
+  function batchRemove (parent, element, node) {
       batch.push(parent.removeChild.bind(parent, element))
-
-      if (oldNode && oldNode.data && oldNode.data.onRemove) {
-        defer(oldNode.data.onRemove, element)
+      if (node.data && node.data.onRemove) {
+        defer(node.data.onRemove, element)
       }
+  }
+
+  function patch (parent, element, oldNode, newNode) {
+    if (oldNode === undefined) {
+      var n = createElementFrom(newNode)
+      parent.insertBefore(n, element)
+      element = n
 
     } else if (
-      node.tag !== oldNode.tag ||
-      typeof node !== typeof oldNode ||
-      typeof node === "string" && node !== oldNode
+      newNode.tag !== oldNode.tag ||
+      typeof newNode !== typeof oldNode ||
+      typeof newNode === "string" && newNode !== oldNode
     ) {
-      if (typeof node === "string") {
-        element.textContent = node
-      } else {
-        var i = createElementFrom(node)
-        parent.replaceChild(i, element)
-        element = i
-      }
-    } else if (node.tag) {
-      updateElementData(element, node.data, oldNode.data)
+      var n = createElementFrom(newNode)
+      parent.replaceChild(n, element)
+      element = n
+
+    } else if (oldNode.tag){
+      updateElementData(element, newNode.data, oldNode.data)
+      patchChildren(element, oldNode, newNode)
+    }
+
+    return element
+  }
+
+  function patchChildren(element, oldNode, newNode) {
+    // We need to index all keyed nodes in oldChildren in order to
+    // distinguish new keyed nodes from nodes to reuse in newChildren
+    var keyedChildren = {}
+    for (var i = 0; i < oldNode.children.length; i++) {
+      if (!oldNode.children[i].data) continue
+      if (!oldNode.children[i].data.key) continue
+      keyedChildren[oldNode.children[i].data.key] = true
+    }
 
-      var len = node.children.length
-      var oldLen = oldNode.children.length
+    var isKeyed = function (node) {
```

Comment:
Mostly for readability. Because they were repeated code snippets. But now that I reduced my 8-branch if-else down to its minimum 5-branches, I have got rid of those extra functions, and just inlined the code in the if-else.